File: frontend/lib/core/models/bill.dart
class Bill {
  final String id;
  final String userId;
  final int month;
  final int year;
  final String planName;
  final double planRate;
  final int totalMeals;
  final int skippedMeals;
  final int messOffMeals;
  final double deductionAmount;
  final double finalAmount;
  final DateTime generatedAt;

  const Bill({
    required this.id,
    required this.userId,
    required this.month,
    required this.year,
    required this.planName,
    required this.planRate,
    required this.totalMeals,
    required this.skippedMeals,
    required this.messOffMeals,
    required this.deductionAmount,
    required this.finalAmount,
    required this.generatedAt,
  });

  factory Bill.fromJson(Map<String, dynamic> json) => Bill(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        month: json['month'] as int,
        year: json['year'] as int,
        planName: json['plan_name'] as String,
        planRate: double.parse(json['plan_rate'].toString()),
        totalMeals: json['total_meals'] as int,
        skippedMeals: json['skipped_meals'] as int,
        messOffMeals: json['mess_off_meals'] as int,
        deductionAmount: double.parse(json['deduction_amount'].toString()),
        finalAmount: double.parse(json['final_amount'].toString()),
        generatedAt: DateTime.parse(json['generated_at'] as String),
      );
}

File: backend/app/schemas/bill.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class BillGenerateRequest(BaseModel):
    month: int = Field(ge=1, le=12)
    year: int = Field(ge=2024, le=2100)


class BillResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    user_full_name: str | None = None
    month: int
    year: int
    plan_name: str
    plan_rate: Decimal
    total_meals: int
    skipped_meals: int
    mess_off_meals: int
    extra_meals_count: int = 0
    extra_meals_amount: Decimal = Decimal("0.00")
    deduction_amount: Decimal
    final_amount: Decimal
    generated_at: datetime
    # The plan's actual seeded monthly rate (never changes) — plan_rate above
    # is the pro-rated amount used for this specific bill's math. Shown
    # alongside start/stop_date so the UI can explain *why* they differ.
    full_monthly_rate: Decimal | None = None
    start_date: date | None = None
    stop_date: date | None = None


class BillGenerateUserRequest(BaseModel):
    user_id: uuid.UUID
    month: int = Field(ge=1, le=12)
    year: int = Field(ge=2024, le=2100)


class BillSummary(BaseModel):
    month: int
    year: int
    total_users: int
    total_revenue: Decimal
    total_deductions: Decimal
